### PR TITLE
Initial Version of Mch Mapping

### PR DESCRIPTION
**Pull request by aphecetche**
Note: as #875 is getting old and is now conflicting with dev (due to O2Dependencies.cmake), I'm opening this new one.
---
This is the initial porting of the MCH mapping from https://github.com/mrrtf/alo/mapping.

This PR introduces :

a header-only library MCHContour (used to compute envelops of e.g. FEE motifs)
a header-only interface library MCHMappingInterface
one implementation MCHMappingImpl3 of the interface above (as the number suggests, there are 2 more in the works, which might -or not- end up in AliceO2 as well at some point)
one library MCHMappingSegContour to help plotting the mapping elements

### CODE REVIEW

**Comment by sawenzel** [comment]
The macos build seems to have a real problem. Do we understand this?

**Comment by aphecetche** [comment]
Seems there is an issue with "include what you use". 
Do I understand that it's a requirement to build our software and that I should have a local version of it installed in order to avoid such issues in the future ?

**Comment by sawenzel** [comment]
I am personally not aware of any "include what you use" usage nor requirement in O2.

**Comment by aphecetche** [comment]
Great. Maybe the macOS build is being a bit overzealous then ?

**Comment by sawenzel** [comment]
@dberzano : Any idea what is the problem here?

**Comment by dberzano** [comment]
I am having a look.

**Comment by dberzano** [comment]
So, on our test Mac, we have IWYU installed, and O2 is enabling it if found. I think our IWYU support is to be blamed. As a workaround, IWYU has been uninstalled from the test machine, but this problem has to be addressed in a different way.

**Comment by dberzano** [comment]
@aphecetche just to clarify: the build wasn't being overzealous. IWYU wasn't triggered at all, it seems there is a problem in the IWYU rules generation for CMake instead.

**Comment by aphecetche** [comment]
The log linked in the comment above (https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/886/d14ed2bd933b89a744cec9b49c8bd229e5471876/fullLog.txt) is again misleading (labelled build/O2/o2-dev-fairroot) but is a macOS one really ?

**Comment by aphecetche** [comment]
@sawenzel @dberzano btw, is it the macOS build failure which is holding the merging of this PR or something else should be attended to ?

**Comment by dberzano** [comment]
Hi @aphecetche, I believe the build directory on macOS is dirty or something like that. Let me have a look. As for the log file names, it's a bug and I'll be taking care of it later.

**Comment by dberzano** [comment]
@aphecetche as said, I was expecting this...

**Comment by sawenzel** [comment]
@dberzano : Following your comments, I am not sure whether we need to fix something before merging this PR or whether we can go ahead?

**Comment by dberzano** [comment]
Hi, a few minutes ago we've done this: alisw/ali-bot#1038. Once merged, you can just wait and the problem will be gone. Note that today not only Sourceforge is down (hence the aforementioned PR), but Travis too.

**Comment by sawenzel** [comment]
@dberzano : Thanks.

**Comment by sawenzel** [comment]
@dberzano : The O2/macos build is still reported to fail but there is no link to the error.
Is it still the IWYU problem or something else?

**Comment by dberzano** [comment]
I think I have kind of fixed the macOS problem. Let's see the result now.

**Comment by dberzano** [comment]
OK @aphecetche, @sawenzel, I can tell you that the build went past CMake, and the problem was that the build directory was dirty. This is definitely an issue of the build system and I'll be taking care of it. Let's wait for the green tick here.

**Comment by dberzano** [comment]
And there we go :-)

**Comment by sawenzel** [comment]
Squash merging for now.

Looking at the polygon code, I am wondering whether such things are not available externally? I am thinking of OpenCV, Boost.polygon for instance ...

**Comment by aphecetche** [comment]
@sawenzel concerning the polygon code, I fully understand the questioning, as I had the same initial one ;-) 

The "raison d'être" of this part of the code is the method to find the contour of a set of (simple) polygons. (createContour function).

While I don't know OpenCV, I did briefly look at Boost.polygon (which I found quite complex and not really suitable for my needs) and then investigated a bit Boost.geometry. The union_ function for instance was promising but was not simplifying the contours enough. So in the end I decided to port the code we used in AliRoot.

**Comment by dberzano** [comment]
@aphecetche @sawenzel for the record, I think I fixed (on Mac; now deploying it elsewhere) the problem of the dirty build directory giving unexpected C.I. results (alisw/ali-bot#544).

File: Detectors/MUON/MCH/Mapping/Interface/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+O2_SETUP(NAME MCHMappingInterface)
+set(BUCKET_NAME mch_mapping_interface_bucket)
+
+add_library(MCHMappingInterface INTERFACE)
+
+install(DIRECTORY include/MCHMappingInterface DESTINATION include)
+
+# custom target so that some IDE (e.g. CLion) recognize those files as
+# being part of the project even though they are not part of a CMake target
+add_custom_target(MCHMappingInterface_ide SOURCES
```
Comment (matthiasrichter) [inline Detectors/MUON/MCH/Mapping/Interface/CMakeLists.txt]: @aphecetche, this is actually the reason for the conflict with the IWYU custom target. While I do not want to exclude an issue in the way, IWYU is integrated, I'm wondering if there is probably a better way to get the visibility of the files in the IDE. This looks very much like a hack, we should find a better solution for.

Comment (aphecetche) [reply Detectors/MUON/MCH/Mapping/Interface/CMakeLists.txt]: @matthiasrichter well, don't know if there's a different way or whether this is a hack, but anyway I don't understand how a custom target can interfere with another completely unrelated one (the iwyu one(s)) ?

Comment (aphecetche) [reply Detectors/MUON/MCH/Mapping/Interface/CMakeLists.txt]: btw, maybe @Barthelemy as a CLion user have some idea here ?

Comment (matthiasrichter) [reply Detectors/MUON/MCH/Mapping/Interface/CMakeLists.txt]: actually I was wrong, these custom_targets do not trigger the issue, it's introduced somewhere else in the PR.

**Pull request merged**